DimensionPreloadEvent

goreacraft opened the pull request:
By canceling this event will …prevent the dimension from being initialized and then you can use

DimensionManager.setWorld to add your own custom dimension instead. This will make possible to create a proper multiverse server side plugin

I added this event here to have the option to use it if you need it as i don't know another way to do it unless using Mixin, and why adding another dependency just for this?

I made this particularly for 1.10.x because i think this version will be used for some time ahead by modpacks as there are lots of mods not updated to 1.11.x yet ( see FTB for ex)

Special care has to be made when players will connect in a dimension that it's initialization is canceled with this event

This is a remake of the commit i made earlier : #3755

goreacraft committed:
Added Cancelable DimensionPreloadEvent. By canceling this event will prevent the dimension from being initialized and then you can use.
DimensionManager.setWorld to add your own custom dimension instead. This will make possible to create a proper multiverse server side plugin

This is a remake of the commit i made earlier : #3755

diesieben07 commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -216,7 +216,7 @@ public static void initDimension(int dim)
         try
         {
-            DimensionManager.getProviderType(dim);
+            DimensionType type = DimensionManager.getProviderType(dim); if(MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.DimensionPreloadEvent(dim, type))) return;
```
This is a Forge file, no need to keep patches small.

diesieben07 commented on src/main/java/net/minecraftforge/event/world/DimensionPreloadEvent.java:
```diff
@@ -0,0 +1,35 @@
+import net.minecraftforge.common.MinecraftForge;
+
+/**
+ * Created by goreacraft on 17/03/03.
```
Author notes are not used in Forge.

goreacraft replied:
oh...ok didn't know that, i thought author notes are there to know who to blame on ;)

goreacraft replied:
can you help me explain what would be the proper way of fixing this now? 
should i add a new commit and remove this one? or there is a way to edit this things on the fly

diesieben07 replied:
Just make a new commit. Github can squash commits when merging your PR, so you do not need to do that yourself.

diesieben07 commented on src/main/java/net/minecraftforge/event/world/DimensionPreloadEvent.java:
```diff
@@ -0,0 +1,35 @@
+ * dimension 0 (Overworld) will not trigger this event
+ * {@link DimensionManager#initDimension(int)},
+ * <br>
+ * This event is {@link Cancelable}.<br>
```
Please explain what happens when you cancel this event.

goreacraft replied:
i will give soon an example of use in more detail with explanation, i hope by tomorrow...
for now you can see what the forgeessentials guys had to do using mixin to cancel function initDimension and make their multiverse module work. Having this event it will not be necessary to use mixin.  (https://github.com/ForgeEssentials/ForgeEssentials/blob/110/develop/src/main/java/com/forgeessentials/core/preloader/mixin/MixinDimensionManager.java )
I am working on a multiverse module myself and i "hit this wall" by not having an event for this, so i fix it by adding it.
I am sorry if i don't know how to make my intentions more clear then this, if you still need more clarifications i will try with the example soon.

goreacraft replied:
this is not an event intended for regular use, it is meant especially for cases like this.
I know that if used wrongly can cause problems, but if used correctly can be a life saver

diesieben07 replied:
This event is a direct port of the Mixin, it seems like. But with an actual Forge edit you could achieve a much better API.

goreacraft replied:
i am not part of forgeessentials dev team, i was looking to see how others found a solution for this problem, and i don't want to use another dependency when it can be done with a simple forge edit

diesieben07 reviewed (commented):
Apart from my individual comments I still do not understand where you are going with this. This event seems very counter-intuitive to me.
If you want to keep this approach, please add detailed Javadocs on how to use it.

goreacraft committed:
Added Cancelable DimensionPreloadEvent. By canceling this event will prevent the dimension from being initialized and then you can use.
DimensionManager.setWorld to add your own custom dimension instead. This will make possible to create a proper multiverse server side plugin

This is a remake of the commit i made earlier : #3755

goreacraft committed:
Added Cancelable DimensionPreloadEvent. By canceling this event will prevent the dimension from being initialized and then you can use.
DimensionManager.setWorld to add your own custom dimension instead. This will make possible to create a proper multiverse server side plugin

This is a remake of the commit i made earlier : #3755

goreacraft committed:
Merge remote-tracking branch 'origin/1.10.x' into 1.10.x

goreacraft committed:
Merge remote-tracking branch 'origin/1.10.x' into 1.10.x

diesieben07 commented on src/main/java/net/minecraftforge/event/world/DimensionPreloadEvent.java:
```diff
@@ -0,0 +1,33 @@
+ * <br>
+ * event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+ * special care should be used when canceling this event
+ * use it only when you know what you are doing
```
That's still not very helpful. If an API is dangerous, it is a bad API, in my opinion. If a certain action must be taken  when this event is cancelled, then enforce that action by not allowing `setCanceled` without that option.

goreacraft replied:
I will try to reformulate and hope i can explain better my resoning: this event it is meant to help with a specific problem as i tried to explain with the "mixin example above". The idea is to prevent the initDimension to turn any non 0 dimension into a WorldServerMulti.class and give the possibility to initiate another Class<? extends WorldServer> instead, so this way it would be possible to fully customize server side only creation of dimensions.
I do not know exactly how the description of this event would be more useful as i do not have a lot of experience in this matter and please if you can help me in making a better description i would be grateful

goreacraft replied:
I cannot give an example of my own yet as i don't have my plugin/mod ready yet but i will give more details on how the FE does this event canceling. On line 418 in https://github.com/ForgeEssentials/ForgeEssentials/blob/110/develop/src/main/java/com/forgeessentials/multiworld/MultiworldManager.java you can see exactly how an event like this can be used for this purpose. I just want to prevent having to use "mixin" when this can be achieved by having a simple event in the right place in forge instead. The point of an API is to make coding easier so this is what i try to help with, by giving a simpler solution to this specific problem.

diesieben07 replied:
I understand that, but the API is terrible.
If you want to allow replacing the world instance, add a setWorld to the event instead of making it cancelable.

goreacraft replied:
Your solution is interesting, i will try to make an implementation for this too. In the mean time can you please tell why is canceling the instantiation of a dimension so dangerous?  i think i am missing something that you see in this case and would be interesting to know about it.

goreacraft committed:
Merge remote-tracking branch 'origin/1.10.x' into 1.10.x

goreacraft committed:
Merge remote-tracking branch 'origin/1.10.x' into 1.10.x

liach commented:
Summoning @luacs1998 for explanations.

goreacraft commented:
I have removed the Cancelable and i added the option to replace the instantiating dimension instead as **diesieben07** sugested

goreacraft committed:
Merge remote-tracking branch 'origin/1.10.x' into 1.10.x

mezz commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -19,38 +19,23 @@
+import org.apache.logging.log4j.Level;
+
+import java.io.File;
+import java.util.*;
```
Please keep the imports similar to how they were before. Make sure your IDE does not automatically rearrange them or collapse them into `.*` groups if they were not that way already.

mezz commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -214,20 +199,25 @@ public static void initDimension(int dim)
         }
         MinecraftServer mcServer = overworld.getMinecraftServer();
-        ISaveHandler savehandler = overworld.getSaveHandler();
+        //ISaveHandler savehandler = overworld.getSaveHandler();
```
You can remove these commented-out lines.

mezz commented on src/main/java/net/minecraftforge/event/world/DimensionPreloadEvent.java:
```diff
@@ -0,0 +1,40 @@
+package net.minecraftforge.event.world;
```
This new file needs a license header

mezz commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -214,20 +199,25 @@ public static void initDimension(int dim)
         {
-            DimensionManager.getProviderType(dim);
+            DimensionType type = DimensionManager.getProviderType(dim);
+            if( dim!=0 ) {
```
Please follow the spacing style used elsewhere in the project, `if (dim != 0)`

mezz reviewed (changes requested):
Code style review

mezz commented:
The 1.10.x branch is not accepting new features like this, only bug fixes. This will have to target 1.11.x.

mezz commented:
You should also create a test mod that shows how this new event will be used.

goreacraft commented:
**mezz** thank you for your review, i have found  your point by point observations very informative and i agree with your reasoning and conclusion.

I will remake this commit for 1.11.x soon, as i think it an important event to have from server side management

goreacraft committed:
synchronizing DimensionPreloadEvent with the 1.11.x in the future

goreacraft committed:
synchronizing DimensionPreloadEvent with the 1.11.x in the future

goreacraft committed:
Merge remote-tracking branch 'origin/1.10.x' into 1.10.x

goreacraft commented:
is there way for me to take down this pull so only the 1.11.x version will remain active? #3761
I don't want to spam here all the changes i will make on 1.10.x fork

williewillus commented:
hit the close button right about here

\>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> vvvvv

goreacraft commented:
duhhh
